File: algorithms/easy/distribute_candies_among_children_i.ex
defmodule Solution do
  @spec distribute_candies(n :: integer, limit :: integer) :: integer
  def distribute_candies(n, limit) do
    dp(%{}, 3, n, limit)
    |> elem(1)
  end

  def dp(mem, chidren, candies, limit) do
    cond do
      candies == 0 ->
        {mem, 1}

      chidren == 0 || limit == 0 ->
        {mem, 0}

      Map.has_key?(mem, {chidren, candies, limit}) ->
        {mem, mem[{chidren, candies, limit}]}

      true ->
        0..min(candies, limit)
        |> Enum.reduce({mem, 0}, fn c, {mem, acc} ->
          {new_mem, ct} = dp(mem, chidren - 1, candies - c, limit)
          {new_mem, acc + ct}
        end)
        |> then(fn {mem, ct} ->
          Map.put(mem, {chidren, candies, limit}, ct)
          {mem, ct}
        end)
    end
  end
end

File: algorithms/easy/distribute_candies_among_children_i.dart
import 'dart:math';

class Solution {
  int distributeCandies(int n, int limit) {
    int ct = 0;
    for (int i = min(n, limit); i >= 0; i--) {
      for (int j = min(n - i, limit); j >= 0; j--) {
        if (n - i - j > limit) {
          break;
        }
        ct++;
      }
    }
    return ct;
  }
}
